stdlib/LinearAlgebra/src/uniformscaling.jl

```diff
@@ -48,6 +48,31 @@ julia> [1 2im 3; 1im 2 3] * I
 """
 const I = UniformScaling(true)
 
+"""
+    (J::UniformScaling)(n::Integer)
+
+Construct a `Diagonal` matrix from a `UniformScaling`.
+
+!!! compat "Julia 1.1"
+     This method is available as of Julia 1.1.
+
+# Examples
+```jldoctest
+julia> I(3)
+3×3 Diagonal{Bool,Array{Bool,1}}:
+  true    ⋅      ⋅
+   ⋅     true    ⋅
+   ⋅      ⋅     true
+
+julia> (0.7*I)(3)
+3×3 Diagonal{Float64,Array{Float64,1}}:
+ 0.7   ⋅    ⋅
+  ⋅   0.7   ⋅
+  ⋅    ⋅   0.7
+```
+"""
+(J::UniformScaling)(n::Integer) = Diagonal(fill(J.λ, n))
```

Comment:
`StepRangeLen(J.λ, 0, n)`, but probably a step too cute.

Reply:
`StepRangeLen` carries redundant data.

There is
```julia
using FillArrays
(J::UniformScaling)(n::Integer) = Diagonal(Fill(J.λ, n))
```
This carries redundant data, too. `typeof(1I(3))` is `Diagonal{Int64,Fill{Int64,1,Tuple{Base.OneTo{Int64}}}}`.

And the big pile of right brackets is kinda scary.

Reply:
We don't have FillArrays in the stdlib.

Reply:
Yes, I was thinking more long term. Of course it can't go in v1.1

Reply:
> Of course an implementation with `Eye` can't go in v1.1

Why not? We can add new features in minor releases. Not sure if we should, but we can.

Reply:
One (probably) crazy idea is to define

```julia
filled(λ, n) =
    Broadcast.Broadcasted{Broadcast.DefaultArrayStyle{1}}(
        identity, (λ,), (Base.OneTo(n),))

(J::UniformScaling)(n::Integer) = Diagonal(filled(J.λ, n))
```

although it does not work at the moment since `Diagonal` needs an abstract vector. Also, `typeof(filled(λ, n))` is very long: `Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1},Tuple{Base.OneTo{Int64}},typeof(identity),Tuple{Float64}}`.  But the nice thing is that this approach lets you have efficient methods like `mul!` for free.

Reply:
> Why not? We can add new features in minor releases. Not sure if we should, but we can.

[1.1 branch date approaching: Dec. 6](https://discourse.julialang.org/t/1-1-branch-date-approaching-dec-6/18233)

It seems to me that using FillArrays would involve committing at the last minute to substantially more new design than the current PR. That's why I conflated "should" with "can".

Reply:
Ah yes, timing-wise 1.1 is too soon. I thought you were saying in terms of compatibility.

Reply:
I expect that `I(n)` will mostly be used in places where performance isn't critical so I'm not sure it's worth being too cute here to avoid the extra allocations. Sure, if we had the functionality of `FillArrays` available then we should use it here but maybe this application isn't sufficient motivation to migrate the functionality to base.

Reply:
My prediction is one can't predict when dense, dense diagonal, or purely structural<sup>1</sup> will be convenient or efficient. Offer a few discoverable options and let the market decide. FillArrays already has `typeof(copy(Eye(n))) == Diagonal{Float64,Array{Float64,1}}` and `typeof(Array(Eye(n))) == Array{Float64,2}`. (I don't know if there are alternatives to `FillArrays`) 

If `Eye(n)` is reliable (i.e. works as a dropin replacement for all the code that uses `eye`), then recommending `use FillArrays` and `Eye(n)` doesn't seem overly burdensome.

Regarding predictions about use patterns and efficiency: I tried representing identity matrices with each of 1) `Eye` (`Diagonal{Float64,Ones{Float64,1,Tuple{Base.OneTo{Int64}}}}`), 2) `Diagonal{Float64,Array{Float64,1}}`, and 3) `Array{Float64,2}` [in an existing package](https://github.com/BBN-Q/QuantumInfo.jl/pull/10#issuecomment-445450354). The result for the test suite (with it's choice of sizes, types, etc.)  is that `Array{Float64,2}` is fastest, both for jit and execution. This includes allocating every time it is used.

<sup>1</sup> "dense", "allocated", "packed"? None are precise.

Reply:
New matrix multiplication API may need the O(1) version of `I(n)`. See: https://github.com/JuliaLang/julia/issues/23919#issuecomment-447216076